driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -258,6 +329,80 @@ private Integer executeBatch(
+                funcCallback -> withAsyncSourceAndConnection(binding::getWriteConnectionSource, true, funcCallback,
+                        (connectionSource, connection, resultCallback) -> {
+                            ConnectionDescription connectionDescription = connection.getDescription();
+                            boolean effectiveRetryWrites = isRetryableWrite(
+                                    retryWritesSetting, effectiveWriteConcern, connectionDescription, sessionContext);
+                            retryState.breakAndThrowIfRetryAnd(() -> !effectiveRetryWrites);
+                            resultAccumulator.onNewServerAddress(connectionDescription.getServerAddress());
+                            retryState.attach(AttachmentKeys.maxWireVersion(), connectionDescription.getMaxWireVersion(),
+                                            true)
+                                    .attach(AttachmentKeys.commandDescriptionSupplier(), () -> BULK_WRITE_COMMAND_NAME,
+                                            false);
+                            ClientBulkWriteCommand bulkWriteCommand = createBulkWriteCommand(
+                                    retryState, effectiveRetryWrites, effectiveWriteConcern, sessionContext,
+                                    unexecutedModels, batchEncoder,
+                                    () -> retryState.attach(AttachmentKeys.retryableCommandFlag(), true, true));
+                            executeBulkWriteCommandAndExhaustOkResponseAsync(
+                                    retryState, connectionSource, connection, bulkWriteCommand, effectiveWriteConcern,
+                                    operationContext,
+                                    resultCallback);
+                        }));
+
+        beginAsync()
+                .<ExhaustiveClientBulkWriteCommandOkResponse>thenSupply(callback -> {
+                    retryingBatchExecutor.get(callback);
+                })
+                .<Integer>thenApply((bulkWriteCommandOkResponse, callback) -> {
+                    callback.complete(resultAccumulator.onBulkWriteCommandOkResponseOrNoResponse(
+                            batchStartModelIndex, bulkWriteCommandOkResponse, batchEncoder.intoEncodedBatchInfo()));
+                }).onErrorIf(throwable -> true, (t, callback) -> {
+                    if (t instanceof MongoWriteConcernWithResponseException) {
+                        callback.complete(resultAccumulator.onBulkWriteCommandOkResponseWithWriteConcernError(
+                                batchStartModelIndex, (MongoWriteConcernWithResponseException) t, batchEncoder.intoEncodedBatchInfo()));
+                    } else if (t instanceof MongoCommandException) {
+                        resultAccumulator.onBulkWriteCommandErrorResponse((MongoCommandException) t);
+                        callback.completeExceptionally(t);
+                    } else if (t instanceof MongoException) {
```

Comment:
Let's rename the corresponding sync exception variable names to `t` to make code more similar.